core/control-panel/impl/src/services/registry.rs

```diff
@@ -48,14 +58,146 @@ impl RegistryService {
+
+        RegistryMapper::fill_from_create_input(&mut entry, &input);
+
+        match &input.value {
+            control_panel_api::RegistryEntryValueInput::WasmModule(module) => {
+                let artifact_id = self.artifact_service.create(module.wasm_module.clone())?;
+
+                entry.value = RegistryValue::WasmModule(WasmModuleRegistryValue {
+                    wasm_artifact_id: artifact_id,
+                    version: module.version.clone(),
+                    dependencies: module
+                        .dependencies
+                        .iter()
+                        .map(|dep| dep.clone().into())
+                        .collect(),
+                });
+            }
+        };
+
+        // There can only be one latest entry, so we need to remove the latest tag from the previous entry if it exists.
+        if entry.tags.contains(&LATEST_TAG.to_string()) {
+            let previous_latest = self.registry_repository.find_ids_where(
+                RegistryWhere::clause()
+                    .and_fullname(&entry.fullname())
+                    .and_tags(vec![LATEST_TAG.to_string()]),
+                None,
+            );
+
+            previous_latest
+                .iter()
+                .filter_map(|id| self.get(id).ok())
+                .for_each(|mut e| {
+                    e.tags.retain(|t| t != LATEST_TAG);
+                    self.registry_repository.insert(e.id, e);
+                });
+        }
+
+        entry.validate()?;
+
+        self.registry_repository.insert(entry.id, entry.clone());
+
+        Ok(entry)
     }
 
     /// Updates the registry entry and returns it.
-    pub fn edit(&self, _registry_id: &RegistryEntryId) -> ServiceResult<RegistryEntry> {
-        unimplemented!()
+    pub fn edit(
+        &self,
+        registry_id: &RegistryEntryId,
+        input: RegistryEntryUpdateInput,
+    ) -> ServiceResult<RegistryEntry> {
+        let mut entry = self.get(registry_id)?;
+
+        RegistryMapper::fill_from_update_input(&mut entry, &input);
+
+        match (&input.value, &entry.value) {
+            (
+                Some(control_panel_api::RegistryEntryValueInput::WasmModule(module)),
+                RegistryValue::WasmModule(current_module),
+            ) => {
+                self.artifact_service
+                    .remove_by_id(&current_module.wasm_artifact_id)?;
+
+                let artifact_id = self.artifact_service.create(module.wasm_module.clone())?;
+
+                entry.value = RegistryValue::WasmModule(WasmModuleRegistryValue {
+                    wasm_artifact_id: artifact_id,
+                    version: module.version.clone(),
+                    dependencies: module
+                        .dependencies
+                        .iter()
+                        .map(|dep| dep.clone().into())
+                        .collect(),
+                });
+            }
+            (None, _) => (),
+        };
+
+        entry.validate()?;
```

Comment:
if this fails then the original entry will stay but its artifact is already deleted, and the new one is dangling.

Reply:
does `latest` tag need check here? maybe it's added by this edit.

Reply:
Great point, just made a change that accounts for that